README.md
<br/>
<h1 align="center">
  Teensy Tiny UUID v4 Generator
</h1>
<p align="middle">
  An RFC 4122 compliant UUID v4 generator compatible with browser and Node APIs.
  </p>
<br/>
<p align="center">
  <a href="https://github.com/tbusillo/teensy-typescript-package/actions/workflows"><img src="https://github.com/tbusillo/teensy-typescript-package/actions/workflows/test.yml/badge.svg" alt="CI status"></a>
</p>
<br/>

## Features

- 3.5kb minified, with no dependencies
- RFC 4122 compliant UUID v4
- Isomorphic - it works in the browser as well as with Node.js, without use of browserify, a bundler, or other dependencies

## Installation

```bash
npm install @teensy/uuid

yarn add @teensy/uuid

pnpm add @teensy/uuid
```

## Usage

```typescript
import uuid from '@teensy/uuid'

const id = uuid() // example output: 8b2e5e9e-2a5b-4e7b-8e1a-4b1d9a4e7f2a
```

## Releasing

If you intend to distribute your package via a manager (e.g., `npm`, `yarn`, etc), this template
includes [standard-version](https://github.com/conventional-changelog/standard-version).

## Releasing

To release a new version, run the following command:

```bash

pnpm run release

```

Canary versions are released automatically on every push to `main` via CI. To use a canary version in your project, use `@teensy/uuid@canary`.

## License

MIT License

Copyright (c) 2023 Tom


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.15](https://github.com/tbusillo/teensy-uuid/compare/v0.0.14...v0.0.15) (2023-06-24)

### [0.0.14](https://github.com/tbusillo/teensy-uuid/compare/v0.0.12...v0.0.14) (2023-06-24)

### Bug Fixes

- remove node scope from crypto import ([#31](https://github.com/tbusillo/teensy-uuid/issues/31)) ([80b43ea](https://github.com/tbusillo/teensy-uuid/commit/80b43eae939f825ad7a0a2895b6c1803fd338b38))

### [0.0.13](https://github.com/tbusillo/teensy-uuid/compare/v0.0.12...v0.0.13) (2023-06-03)

### [0.0.12](https://github.com/tbusillo/teensy-uuid/compare/v0.0.11...v0.0.12) (2023-06-02)

### [0.0.11](https://github.com/tbusillo/teensy-uuid/compare/v0.0.10...v0.0.11) (2023-06-02)

### [0.0.10](https://github.com/tbusillo/teensy-uuid/compare/v0.0.9...v0.0.10) (2023-04-02)

### Bug Fixes

- update pnpm lock file ([1ee3c89](https://github.com/tbusillo/teensy-uuid/commit/1ee3c890c772fed87439bbe146ebf8c60a9ab7a1))

### [0.0.9](https://github.com/tbusillo/teensy-uuid/compare/v0.0.8...v0.0.9) (2022-12-28)

### Bug Fixes

- make package.json private equal to false ([5e82580](https://github.com/tbusillo/teensy-uuid/commit/5e8258081e7a4abb711bba46ac7f6577cb164b45))
- update publish config ([1100183](https://github.com/tbusillo/teensy-uuid/commit/1100183f5914636fdd4f95db2fc705620bfe2212))

### [0.0.8](https://github.com/tbusillo/teensy-uuid/compare/v0.0.7...v0.0.8) (2022-12-03)

### Features

- browser and node working ([72fec1a](https://github.com/tbusillo/teensy-uuid/commit/72fec1aa537e4120bc5c38d1d851927ee0f6dffb))

### [0.0.7](https://github.com/tbusillo/teensy-uuid/compare/v0.0.6...v0.0.7) (2022-11-29)

### [0.0.6](https://github.com/tbusillo/teensy-uuid/compare/v0.0.5...v0.0.6) (2022-11-27)

### [0.0.5](https://github.com/tbusillo/teensy-uuid/compare/v0.0.4...v0.0.5) (2022-11-27)

### Bug Fixes

- make crypto use webcrypto ([5783319](https://github.com/tbusillo/teensy-uuid/commit/5783319779085818044486fd9e7c621f090a9e07))

### [0.0.4](https://github.com/tbusillo/teensy-uuid/compare/v0.0.3...v0.0.4) (2022-11-27)

### Bug Fixes

- add ternary to check for presence of crypto ([178aeb1](https://github.com/tbusillo/teensy-uuid/commit/178aeb10f632a80017e7a85143f5ffb22cb33e77))

### [0.0.3](https://github.com/tbusillo/teensy-uuid/compare/v0.0.2...v0.0.3) (2022-11-27)

### Bug Fixes

- make build generate correct output ([9fe84cd](https://github.com/tbusillo/teensy-uuid/commit/9fe84cdbfd1a3103cf6418fc8295a0bc686ebafc))

### [0.0.2](https://github.com/tbusillo/teensy-uuid/compare/v0.0.1...v0.0.2) (2022-11-27)

### Features

- adds error handling to package ([2c618d8](https://github.com/tbusillo/teensy-uuid/commit/2c618d8fdd33a541869351209190db9ff4884ba4))

### Bug Fixes

- release related changes missed in release ([6f4374e](https://github.com/tbusillo/teensy-uuid/commit/6f4374e7fd40d4ead30f67ff4d0a170dc5617c23))

### 0.0.1 (2022-11-27)

### Features

- initial working package and test coverage ([a14e68a](https://github.com/tbusillo/teensy-uuid/commit/a14e68aa95b006b6e5e100798c971275c6eddf87))


test/index.test.ts
import { describe, it, assert } from 'vitest'
import uuid from '../src/index'

describe('uuid', () => {
  it('returns a UUID without errors', () => {
    assert.typeOf(uuid(), 'string')
  })

  it('returns a UUID in the correct format', () => {
    const id = uuid()
    const pattern = new RegExp(
      /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i
    )
    assert.match(id, pattern)
    assert.equal(id.length, 36)
  })

  it('generates unique IDs', () => {
    const id1 = uuid()
    const id2 = uuid()

    assert.notEqual(id1, id2)
  })

  describe('when used in a browser', () => {
    it('returns a UUID without errors', () => {
      assert.typeOf(uuid(), 'string')
    })
  })

  describe('when passed a config of `short=true`', () => {
    it('returns a short UUID without errors', () => {
      const result = uuid({ short: true })
      assert.typeOf(result, 'string')
      assert.lengthOf(result, 6)
    })
  })
})


src/tests/index.test.ts
import { describe, it, assert, vi, afterEach, beforeEach } from 'vitest'
import uuid, { browser } from '../index'

describe('uuid', () => {
  it('returns a UUID without errors', () => {
    assert.typeOf(uuid(), 'string')
  })

  it('returns a UUID in the correct format', () => {
    const id = uuid()
    const pattern = new RegExp(
      /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i
    )
    assert.match(id, pattern)
    assert.equal(id.length, 36)
  })

  it('generates unique IDs', () => {
    const id1 = uuid()
    const id2 = uuid()

    assert.notEqual(id1, id2)
  })

  describe('when used in a browser', () => {
    it('returns a UUID without errors', () => {
      assert.typeOf(uuid(), 'string')
    })
  })

  describe('browser', () => {
    let _crypto: Crypto = globalThis.crypto

    beforeEach(() => {
      _crypto = globalThis.crypto
    })

    afterEach(() => {
      globalThis.crypto = _crypto
    })

    it('throws an error when globalThis.crypto is not available', () => {
      const crypto = globalThis.crypto

      vi.stubGlobal('crypto', undefined)

      assert.throws(() => {
        browser()
      })
      globalThis.crypto = crypto
    })
  })
})


src/tests/browser.spec.ts
import { test, expect } from '@playwright/test'

test('has title', async ({ page }) => {
  await page.goto('http://localhost:3005')

  // Expect a title "to contain" a substring.
  await expect(page).toHaveTitle(
    /@teensy\/uuid browser devleopment\/test fixture/
  )
})

test('shows uuid', async ({ page }) => {
  await page.goto('http://localhost:3005')
  const element = page.getByText(/uuid:/i)
  await expect(element).toHaveText(/uuid: [0-9a-f-]{36}/i)
})


src/node.ts
import { formatUUID, setSignificantBits } from './utils'

/**
 * @description Generates a 128 bit randomly-generated and RFC 4122 version 4 UUID
 *
 * @returns {string} Output a 36-character string in the folowing format: "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX"
 */

export const node = (): string => {
  const crypto = require('crypto').webcrypto
  // Generate 16 random bytes
  const bytes: Uint8Array = crypto.getRandomValues(new Uint8Array(16))

  if (!bytes || bytes.length !== 16) {
    throw new Error(
      'Unable to generate random bytes using crypto.getRandomValues()'
    )
  }

  const encodedBytes = setSignificantBits(bytes)

  // Encode the adjusted bytes as 32 hexadecimal digits
  const hex: string = Buffer.from(encodedBytes).toString('hex')

  if (typeof hex !== 'string' || !hex || hex.length !== 32) {
    throw new Error(
      'Unable to encode random bytes as a string of hexadecimal digits'
    )
  }

  // Add four hyphen "-" characters to obtain blocks of 8, 4, 4, 4 and 12 hex digits
  const result = formatUUID(hex)

  if (!result) {
    throw new Error('Unable to generate a UUID')
  }

  return result
}

export default node


src/utils.ts
const convertToHex = (vals: Uint8Array): string => {
  if (typeof window !== 'undefined') {
    return [...vals].map(val => val.toString(16).padStart(2, '0')).join('')
  }
  return Buffer.from(vals).toString('hex')
}

const formatUUID = (hex: string): string => {
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(
    12,
    16
  )}-${hex.slice(16, 20)}-${hex.slice(20)}`
}

const setSignificantBits = (bytes: Uint8Array): Uint8Array => {
  // set the four most significant bits of the 7th byte to 0100'B, so the high nibble is "4"
  bytes[6] = (bytes[6] & 0x0f) | 0x40
  // set the two most significant bits of the 9th byte to 10'B, so the high nibble will be one of "8", "9", "A", or "B"
  bytes[8] = (bytes[8] & 0x3f) | 0x80
  return bytes
}

export { convertToHex, formatUUID, setSignificantBits }


build.config.ts
import { defineBuildConfig } from 'unbuild'

export default defineBuildConfig({
  // If entries is not provided, will be automatically inferred from package.json
  entries: [
    // default
    './src/index',

    // mkdist builder transpiles file-to-file keeping original sources structure
    // {
    //   builder: 'rollup',
    //   emitCjs: true,
    //   input: './src',
    //   outDir: './dist',
    //   format: 'esm',
    //   ext: 'mjs'
    // }
    {
      builder: 'rollup',
      input: './src/index',
      outDir: './dist',
      format: 'esm'
    }
  ],
  rollup: {
    // If you want to use rollup plugins, you can use this option
    emitCJS: true
  },
  declaration: true,
  // Change outDir, default is 'dist'
  outDir: 'dist',

  // Generates .d.ts declaration file
  failOnWarn: false
})


src/browser.ts
import { convertToHex, formatUUID, setSignificantBits } from './utils'

/**
 * @description Generates a 128 bit randomly-generated and RFC 4122 version 4 UUID
 *
 * @returns {string} Output a 36-character string in the folowing format: "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX"
 */

export const browser = (): string => {
  const crypto: Crypto =
    !!(globalThis && globalThis.crypto && globalThis.window) &&
    globalThis.crypto

  if (!crypto) {
    throw new Error(
      'globalThis.crypto is not available. Are you in a Node environment?'
    )
  }

  const { getRandomValues } = crypto as Crypto

  if (getRandomValues) {
    // Generate 16 random bytes
    const bytes: Uint8Array = getRandomValues(new Uint8Array(16))

    if (!bytes || bytes.length !== 16) {
      throw new Error(
        'Unable to generate random bytes using crypto.getRandomValues()'
      )
    }

    const encodedBytes = setSignificantBits(bytes)

    // Encode the adjusted bytes as 32 hexadecimal digits
    const hex = convertToHex(encodedBytes)

    if (!hex || typeof hex !== 'string' || hex.length !== 32) {
      throw new Error(
        'Unable to encode random bytes as a string of hexadecimal digits'
      )
    }

    // Add four hyphen "-" characters to obtain blocks of 8, 4, 4, 4 and 12 hex digits
    const uuid = formatUUID(hex)

    if (!uuid) {
      throw new Error('Unable to generate a UUID')
    }

    return uuid
  }
  throw new Error('globalThis.crypto is not available')
}
